Add distraction free to site editor

draganescu opened the pull request:
<!-- Thanks for contributing to Gutenberg! Please follow the Gutenberg Contributing Guidelines:
https://github.com/WordPress/gutenberg/blob/trunk/CONTRIBUTING.md -->

## What?
<!-- In a few words, what is the PR actually doing? -->

Closes #49175
This PR brings over distraction free mode to the site editor.

## Why?
<!-- Why is this PR necessary? What problem is it solving? Reference any existing previous issue(s) or PR(s), but please add a short summary here, too -->

- Because distraction free work is a good environment to cultivate
- Because the command center makes full chrome UI useless for power users
- Because it enables a really 1:1 preview - while maintaining everything editable

## How?
<!-- How is your PR addressing the issue at hand? What are the implementation details? -->

1. Adds a setting to the more menu of the site editor
2. Look for distraction free being off for displaying some chrome

A part of the work just works as the block editor and interface packages already know about this setting from the previous work on the post editor.

## Testing Instructions
<!-- Please include step by step instructions on how to test this PR. -->
<!-- 1. Open a Post or Page. -->
<!-- 2. Insert a Heading Block. -->
<!-- 3. etc. -->

1. Have a block theme active
4. Go to Appearance -> Editor
5. Click on the preview frame
6. From the top right dotted menu choose DIstraction Free
7. Test it out

### Testing Instructions for Keyboard
<!-- How can you test the changes by using the keyboard only? Please note, this is required for PRs that change the user interface (UI). This ensures the PR can be tested for any possible accessibility regressions. -->

N/A

## Screenshots or screencast <!-- if applicable -->


https://github.com/WordPress/gutenberg/assets/107534/e0d7afd9-1ea0-4e79-a486-3d027fa6c4cc


## To do

- [ ] Add distraction free to the preferences modal to the site editor
- [ ] Add distraction free keyboard shortcut to the site editor
- [x] Add animation to parts of the header

draganescu committed:
adds preference setter in UI

Commit:
Fix distraction free toolbar showing

The store has the variable `distractionFree`, but `isDistractionFree` was incorrectly being used.

draganescu committed:
fix distraction free flag in site editor, revert change in block toolbar hook

draganescu committed:
basic distraction free enabled

draganescu committed:
hide header items in distraction free

draganescu commented:
At this stage this PR needs to receive some design guidance. I am unsure that copying the behavior of the setting from the post editor is enough. Having some hands on interaction could explain better my concerns in #49175,

alexstine commented:
@draganescu I also think this PR should not move forward until you find a fix for the reported accessibility issue that still exists in the post editor with total focus loss after toggling the setting.

draganescu commented:
> I also think this PR should not move forward until you find a fix for the reported accessibility issue that still exists in the post editor with total focus loss after toggling the setting.

@alexstine yes, absolutely. I still do not have a solution there for now.

draganescu committed:
fix the JS test for settings

draganescu committed:
fix the other test for editor settings

draganescu committed:
remove the rerendering of the site editor header to maintain focus on the writing options menu item

draganescu committed:
animated hader in edit mode for distraction free

draganescu committed:
fix sidebar header

getdave commented on packages/edit-site/src/components/header-edit-mode/index.js:
```diff
@@ -156,6 +162,11 @@ export default function HeaderEditMode() {
 		window?.__experimentalEnableZoomedOutView && isVisualMode;
 	const isZoomedOutView = blockEditorMode === 'zoom-out';
 
+	const slideY = {
+		hidden: { y: '-50px' },
+		hover: { y: 0, transition: { type: 'tween', delay: 0.2 } },
+	};
```
If this object is non-dynamic, then you could put it outside the component to get the benefits of referential equality.

draganescu replied:
yea good call, I think this is the case in other places as well

draganescu committed:
move static objects out of components

jeryj committed:
Refactor distractionFree site editor animations

All animations are now triggered by the edit-site-layout__header-container which has an animationState determined by the view and isDistractionFree mode. It also currently sets a isDistractionFreeHovering animation when hovering that element. This will likely need to change to be able to keep the distraction free header visible when one of the header popovers is open.

jeryj committed:
Remove invalid animation values for initial and exit

jeryj committed:
Fix bug where edit-mode__actions would get stuck in isDistractionFreeState

The actions would get stuck in the isDistractionFree state when:
- loading the page in the editor view with distraction free on
- turn off distraction free
- action stuck above the frame

Why? Because the initial state of that component was -50px on page load, so that's what it would transition to when isDistractionFree was turned off. To fix this, we needed to either set an initial state of x: 0 so it can transition to that property, OR assign values for the other states like view or edit so it has a place to transition to.

jeryj committed:
Override animation with !important CSS when focus is within header on distraction-free mode

jeryj committed:
Adjust animation exit delay for header so the header remains visible longer when hovering out.

draganescu committed:
remove hardcoded z-index

draganescu committed:
turn off distraction free mode when opening the styles sidebar from site view

draganescu commented:
In themes that do not support style variations the styles link in the sidebar opens the edit mode view with the global styles sidebar open. This conflicts with distraction free mode (as the header is always on top of the sidebars it's impossible to close the sidebar) - so I chose to deactivate distraction free automatically when this situation is met.

draganescu committed:
add distraction free mode to preferences panel in the site editor

draganescu committed:
Adds the distraction free mode keyboard shortcut

draganescu commented:
The command center and distrction free are quite nice together:


https://github.com/WordPress/gutenberg/assets/107534/76597135-40b8-4388-8c07-c96f4a33fefc

ntsekouras commented on packages/edit-site/src/components/sidebar-navigation-screen-global-styles/index.js:
```diff
@@ -53,9 +55,19 @@ export function SidebarNavigationItemGlobalStyles( props ) {
 			onClick={ () => {
-				// switch to edit mode.
+				// Disable distraction free mode.
+				toggleFeature( 'distractionFree', false );
```
Why would we need to always toggle this feature? Additionally, this function doesn't accept a second argument, so the `false` here does nothing.

I'm going to remove this as part of https://github.com/WordPress/gutenberg/issues/50924, but let me know if we want it for some reason and should just make it work properly.

draganescu replied:
No no, while the false argument comes from who knows what, the  the mode has to be turned off because it opens the sidebar with the global styles inspector for themes that do not have style variations. That is unexpected for DFM which expects there are no sidebars - because of z-index you can't close sidebars.

So the disabling is intentional at least until we figure out better UX for DFM and the editor chrome.

ntsekouras replied:
@draganescu can you give a quick testing [here](https://github.com/WordPress/gutenberg/pull/51964) that the DFM behavior is the wanted one, when we click to open styles? Thanks!

draganescu replied:
Yes, accesing the edit styles link for themes that don't have styles disables DFM.

afercia commented:
Not sure why but it appears the `SoteHub` now renders an `as="[object Object]"` HTML attribute in the DOM. Maybe because it uses `as={ motion.div }` but it is a `motion.div` already?

<img width="983" alt="Screenshot 2023-07-03 at 10 37 08" src="https://github.com/WordPress/gutenberg/assets/1682452/5f2ec9a9-b320-4d3b-a28b-f137e6969a20">

Cc @draganescu
